bigtable/internal/transport/connpool.go

```diff
@@ -16,53 +16,244 @@ package internal
+	}
+}
+
+// WithMeterProvider provides the meter provider for writing metrics
+func WithMeterProvider(mp metric.MeterProvider) BigtableChannelPoolOption {
+	return func(p *BigtableChannelPool) {
+		p.meterProvider = mp
+	}
+}
+
+// WithLogger provides the logger for logging events
+func WithLogger(logger *log.Logger) BigtableChannelPoolOption {
+	return func(p *BigtableChannelPool) {
+		p.logger = logger
+	}
+}
+
+// WithInstanceName provides the full instance Name
+func WithInstanceName(instanceName string) BigtableChannelPoolOption {
+	return func(p *BigtableChannelPool) {
+		p.instanceName = instanceName
+	}
+}
+
+// WithFeatureFlagsMetadata provides the feature flags metadata
+func WithFeatureFlagsMetadata(featureFlagsMd metadata.MD) BigtableChannelPoolOption {
+	return func(p *BigtableChannelPool) {
+		p.featureFlagsMD = featureFlagsMd
+	}
+}
+
+const (
+	primeRPCTimeout = 10 * time.Second
+)
+
 var errNoConnections = fmt.Errorf("bigtable_connpool: no connections available in the pool")
 var _ gtransport.ConnPool = &BigtableChannelPool{}
 
+// BigtableConn wraps grpc.ClientConn to add Bigtable specific methods.
+type BigtableConn struct {
+	*grpc.ClientConn
+	isALTSConn atomic.Bool
+}
+
+// Prime sends a PingAndWarm request to warm up the connection.
+func (bc *BigtableConn) Prime(ctx context.Context, fullInstanceName, appProfileID string, featureFlagsMd metadata.MD) error {
+	client := btpb.NewBigtableClient(bc.ClientConn)
+	req := &btpb.PingAndWarmRequest{
+		Name:         fullInstanceName,
+		AppProfileId: appProfileID,
+	}
+
+	requestParamsMD := metadata.Pairs(requestParamsHeader,
+		fmt.Sprintf("name=%s&app_profile_id=%s", url.QueryEscape(fullInstanceName), url.QueryEscape(appProfileID)))
+
+	originalContextMd, _ := metadata.FromOutgoingContext(ctx)
+	ctx = metadata.NewOutgoingContext(ctx, metadata.Join(originalContextMd, requestParamsMD, featureFlagsMd))
+
+	// Use a timeout for the prime operation
+	primeCtx, cancel := context.WithTimeout(ctx, primeRPCTimeout)
+	defer cancel()
+
+	var p peer.Peer
+	_, err := client.PingAndWarm(primeCtx, req, grpc.Peer(&p))
+	if err != nil {
+		return err
+	}
+
+	if p.AuthInfo != nil {
+		if _, ok := p.AuthInfo.(alts.AuthInfo); ok {
+			bc.isALTSConn.Store(true)
+		}
+	}
+	return nil
+}
+
+// NewBigtableConn creates a wrapped grpc Client Conn
+func NewBigtableConn(conn *grpc.ClientConn) *BigtableConn {
+	return &BigtableConn{
+		ClientConn: conn,
+	}
+}
+
+// connEntry represents a single connection in the pool.
+type connEntry struct {
+	conn          *BigtableConn
+	unaryLoad     atomic.Int32 // In-flight unary requests
+	streamingLoad atomic.Int32 // In-flight streaming requests
+	errorCount    atomic.Int64 // Errors since the last metric report
+	drainingState atomic.Bool  // True if the connection is being gracefully drained.
+
+}
+
+// isALTSUsed reports whether the connection is using ALTS aka Direct Access.
+// best effort basis
+func (e *connEntry) isALTSUsed() bool {
+	if e.conn == nil {
```

Comment:
please add a comment noting when e.conn is nil

Reply:
I will address this in another PR